Movement Updates

BeksOmega opened the pull request:
<!--
  - Thanks for submitting code to Blockly!  Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## The basics

<!-- TODO: Verify the following, checking each box with an 'x' between the brackets: [x] -->

- [X] I branched from develop
- [X] My pull request is against develop
- [X] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

## The details
### Resolves

<!-- TODO: What Github issue does this resolve? Please include a link. -->
#2205 

### Proposed Changes

<!-- TODO: Describe what this Pull Request does.  Include screenshots if applicable. -->

Adds a new "move" object configuration option with the following properties:
* scrollbars: Determines if the workspace has scrollbars. Defaults to true if the workspace has categories.
* drag: Determines if the workspace can be dragged with the mouse. Always false if scrollbars is false (at least in options parsing). Defaults to true if scrollbars is true.
* wheel: Determines if the workspace can be scrolled with the mouse wheel. Always false if scrollbars is false (at least in options parsing). Defaults to false.

Parsing does not currently allow these options to be independent, but you can make them independent by overriding the drag & wheel options after the workspace has been injected. This behavior is supported.

Changes the zoom function so that the zoom is "anchored" to the mouse position, even without scrollbars.

### Reason for Changes

<!--TODO: Explain why these changes should be made.  Include screenshots if applicable. -->
People have requested scroll on wheel. #110

Also I think it may be useful to start separating the movement functionality from scrollbars to work towards #2145

### Test Coverage

<!-- TODO: Please show how you have added tests to cover your changes,
  -        or tell us how you tested it. For each systems you tested,
  -        uncomment the systems in the list below.
  -->
  
Independent-ness of Scrollbars/Drag/Scrollwheel
1. Set move > scrollbars option to true. Set move > drag option to false. Set move > wheel option to false.
2. Observed that scrollbars alone worked. (pass)
3. Set move > scrollbars option to false. Overwrote move > drag to be true. Set move > wheel option to false. 
4. Observed that the dragging alone worked. And that scrollbars were not created. (pass)
5. Set move > scrollbars option to false. Set move > drag option to false. Overwrote move > wheel to be true.
6. Placed a block so I could observe the scrolling.
7. Observed that the scrollwheel alone worked. And that scrollbars were not created. (pass)

Scrolling and Zooming Interaction
1. Set move > scrollbars option to true. Set move > wheel option to false. Set zoom > wheel option to true.
2. Scrolled the mouse wheel. Observed how the workspace zoomed. (pass)
3. Set move > scrollbars option to true. Set move > wheel option to true. Set zoom > wheel option to true.
4. Scrolled the mouse wheel. Observed how the workspace scrolled vertical. (pass)
5. Hit shift and scrolled the mouse wheel. Observed how the workspace scrolled horizontally. (pass)
6. Hit ctrl and scrolled the mouse wheel. Observed how the workspace zoomed. (pass)
7. Set move > scrollbars option to true. Set move > wheel option to true. Set zoom > wheel option to false.
8. Scrolled the mouse wheel. Observed how the workspace scrolled. (pass)
9. Hit shift and scrolled the mouse wheel. Observed how the workspace scrolled horizontally. (pass)

Zooming With and Without Scrollbars
1. Set scrollbars option to true. Set zoom > wheel option to true. Placed a block to visualize zooming.
2. Scrolled the mouse wheel. Observed how the zoom was anchored to the mouse. (pass)
3. Pressed the zoom in control. Observed how it zoomed towards the center. (pass)
4. Changed scrollbars option to false.
5. Scrolled the mouse wheel. Observed how the zoom was anchored to the mouse. (pass)
6. Pressed the zoom in control. Observed how it zoomed towards the center. (pass)
7. Changed the zoom > wheel option to false. 
8. Pressed the zoom in control. Observed how it zoomed towards the top left corner. (pass)

Note: Steps 7&8 of the above make it so the content is no longer bounded. If you were to, for example, override drag to make the content bounded it would zoom towards the center.

Tested on:
* Desktop Chrome
<!-- * Desktop Firefox -->
<!-- * Desktop Safari -->
<!-- * Desktop Opera -->
<!-- * Windows Internet Explorer 10 -->
<!-- * Windows Internet Explorer 11 -->
<!-- * Windows Edge -->

<!--
* Smartphone/Tablet/Chromebook (please complete the following information):
  * Device: [e.g. iPhone6]
  * OS: [e.g. iOS8.1]
  * Browser [e.g. stock browser, safari]
  * Version [e.g. 22]
-->

### Additional Information

<!-- Anything else we should know? -->
All of this works, but there are still some things I want to do (like clean up calls to scrollbars.set) so I consider it a little bit WIP. I just wanted to post a PR before I went too crazy doing things, so I can still change stuff easily.

jogibear9988 commented on core/inject.js:
```diff
@@ -232,19 +232,19 @@ Blockly.createMainWorkspace_ = function(svg, options, blockDragSurface,
+            metrics.viewHeight + metrics.viewTop ||
             metrics.contentLeft <
-                (options.RTL ? metrics.viewLeft : edgeLeft) ||
+                (options.RTL ? metrics.viewLeft : metrics.viewLeft) ||
```
why don't remove this condition?

BeksOmega replied:
Yep, I need to take another look at this whole function. Coordinates man :/

Thanks for pointing it out to me!

BeksOmega replied:
Turns out that I thought the "+ metrics.absoluteLeft" and "+ metrics.absoluteTop" were causing a problem, but the problem was actually that the scrollX and scrollY were not getting set after setMetrics was called, so viewTop and viewLeft were getting calculated incorrectly. This has been fixed.

Everything (except the mainworkspace.isContentBounded_() checks) has been reverted, and the workspaceChanged function is working once again.

rachel-fenichel commented:
Thanks for sending this in for review before diving deeper.  The scrollbar code has some important performance constraints, and has been optimized in some non-obvious ways to improve performance.  As you play with it you'll want to make sure your changes don't significantly change performance.

The most important thing is that `getMetrics()` isn't free, and we want to use cached values whenever possible.
GetMetrics also gets called in places you wouldn't expect, so it's worth digging through code to make sure they aren't being called unnecessarily.

The screenshot below is from the [performance](https://developers.google.com/web/tools/chrome-devtools/evaluate-performance/reference) tab in Chrome developer tools.  These tools are really helpful for keeping an eye on performance.  For this test I 
- loaded the playground
- hit airstrike once
- used the mouse wheel to scroll in

Here's what that looks like on master, for reference:
![image](https://user-images.githubusercontent.com/13686399/52149590-1257d380-2622-11e9-81d7-d426815e1ca3.png)

So far what you've got looks reasonable.  Just make sure you keep this stuff in mind.

rachel-fenichel commented:
Tag me on this when you're ready to have it reviewed again and I'll take another look.

BeksOmega commented:
Ok @rachel-fenichel I think it's ready for another look!

Pre Script : You know how much I value my PRs being easy to review, but I fell below my own standards on this one. I'll try my best to stick to the 1 PR 1 issue principle in the future, but I really wanted to leave the code in a good place, and that required more code than I first assumed.

### Updates!

The bulk of the changes went into making scrollbar-less zoomable workspaces work better. For example...

Old Zooming With Buttons:
![oldzoom_buttons](https://user-images.githubusercontent.com/25440652/52425334-a1486e00-2ab0-11e9-9bcf-5c9e00bcf682.gif)
New Zooming With Buttons:
![newzoom_buttons](https://user-images.githubusercontent.com/25440652/52425361-af968a00-2ab0-11e9-9c36-45dfd998224d.gif)

Old Zooming With Wheel:
![oldzoom_wheel](https://user-images.githubusercontent.com/25440652/52425405-c63ce100-2ab0-11e9-83be-e22bcbad8922.gif)
New Zooming With Wheel:
![newzoom_wheel](https://user-images.githubusercontent.com/25440652/52425428-d0f77600-2ab0-11e9-92e7-6f2cb33845b0.gif)

Blocks are now completely bumped into the workspace.
Old Block Bumping:
![oldbumping_blocks](https://user-images.githubusercontent.com/25440652/52425850-b1ad1880-2ab1-11e9-9412-1b0ce62848fc.gif)
New Block Bumping:
![newbumping_blocks](https://user-images.githubusercontent.com/25440652/52425867-b8d42680-2ab1-11e9-8b96-ab7bf59258df.gif)

Workspace comments are also bumped now.
Old Comment Bumping:
![oldbumping_comments](https://user-images.githubusercontent.com/25440652/52425927-dacda900-2ab1-11e9-9ae7-78144c1c29f6.gif)
New Comment Bumping:
![newbumping_comments](https://user-images.githubusercontent.com/25440652/52425934-df925d00-2ab1-11e9-8cdd-ecd8d0364469.gif)

The scroll center zoom button was removed for un-movable workspaces (as you can see in the two "new" gifs above), because it could cause you to lose blocks at the edges of the workspace. 

The highlight definition button was also removed for un-movable workspaces, because it could cause you to lose blocks at the edges of the workspace.

All the movement functions now work correctly as well:

Fixed Issue #989 (incorrect scrollCenter() with flyouts)
![newscrollcenter_1](https://user-images.githubusercontent.com/25440652/52426156-4b74c580-2ab2-11e9-9dee-757027b4b83a.jpg)
![newscrollcenter_2](https://user-images.githubusercontent.com/25440652/52426166-4fa0e300-2ab2-11e9-95f9-363179b216b3.jpg)

Fixed zoomToFit with flyouts.
Old zoomToFit:
![oldzoomtofit_1](https://user-images.githubusercontent.com/25440652/52426226-6b0bee00-2ab2-11e9-8b98-597a3cd3a4a1.jpg)
![oldzoomtofit_2](https://user-images.githubusercontent.com/25440652/52426232-6e9f7500-2ab2-11e9-9517-a83a866a7255.jpg)
New zoomToFit:
![newzoomtofit_1](https://user-images.githubusercontent.com/25440652/52426247-752dec80-2ab2-11e9-956b-82cabbe25d96.jpg)
![newzoomtofit_2](https://user-images.githubusercontent.com/25440652/52426256-77904680-2ab2-11e9-859b-888cfb2f1cb8.jpg)

Also fixed issue #944 (zoom to fit incorrect with negative block coords)
![newzoomtofit_negativecoords](https://user-images.githubusercontent.com/25440652/52426309-9262bb00-2ab2-11e9-970b-a7f970f3509a.gif)

Post Script: I'm going to have to do 2000lbs of testing on this before I'm confident that all the edge cases are solid, but I wanted to give you another chance to look at it, and tell me what I should separate out if need be.

Post Post Script: Speaking of testing, if you have any ideas about how I could add unit tests for this I would be very interested in doing so.

Thanks for taking another look!
Beka

BeksOmega commented on core/inject.js:
```diff
@@ -232,71 +232,162 @@ Blockly.createMainWorkspace_ = function(svg, options, blockDragSurface,
+  // A reminder for developers to add any events that have to do with
+  // objects on/in the workspace to the bumpObjects function. Useful when
+  // testing often happens on a bounded workspace.
+  var bumpObjectsEventChecker = function(e) {
```
Do you know of any way to make closure not add this code to the minified files? Or to have it not run in compiled mode?

I added it because people will definitely forget to update this function, especially when most of the testing happens on a moveable workspace. But developers using blockly don't need to see this message.

rachel-fenichel replied:
I don't think we need this function.  Adding events isn't a super common thing yet, and this is the sort of information that we should include in the documentation on how to create a new event type instead of having it be a runtime check.

Rather than checking directly against the event types in your change handler, you can make a list in Blockly.Events of event types that should be bumped, and check the event type against that.  That way you don't have to enumerate all of the event types all over the place, and that code will be easier to read.

You can put a comment on that list saying that anyone adding a new event type that will cause bumping should add it to the list.

BeksOmega commented:
This also fixes #2160:
![programmaticresize](https://user-images.githubusercontent.com/25440652/52524005-b78b3100-2c4c-11e9-8ac6-cdd1ab6a28dd.gif)

There are probably a couple of things that have been fixed because of the reorganization.

rachel-fenichel commented:
Thanks for the PR.  I'm a bit backlogged on reviews because I've been getting a release ready, but I'll give this a full review tomorrow.

BeksOmega commented on core/workspace_svg.js:
```diff
@@ -1516,36 +1628,39 @@ Blockly.WorkspaceSvg.prototype.setBrowserFocus = function() {
-  if (this.scale == newScale) {
-    return;  // No change in zoom.
-  }
-  if (this.scrollbar) {
```
This check encapsulated centering the zoom on the mouse pointer, which was only possible before if there were scrollbars. Now we can center on the mouse pointer no matter what, so the check is unnecessary.

rachel-fenichel commented on core/inject.js:
```diff
@@ -232,71 +232,163 @@ Blockly.createMainWorkspace_ = function(svg, options, blockDragSurface,
+    var bumpObjects = function(e) {
+      // We always check isMovable_ again because the original
+      // "not movable" state of isMovable_ could have been changed.
+      if (!mainWorkspace.isDragging() && !mainWorkspace.isMovable_() &&
```
See my comment on your bump object event checker, below.

rachel-fenichel commented on core/inject.js:
```diff
@@ -232,71 +232,163 @@ Blockly.createMainWorkspace_ = function(svg, options, blockDragSurface,
+  if (!options.readOnly && !mainWorkspace.isMovable_()) {
+    // Helper functions for the workspaceChanged callback.
+    var getWorkspaceMetrics = function() {
+      var workspaceMetrics = Object.create(null);
```
This looks like it's doing a lot of work and checks that normally happen in `getMetrics`.  Why is all the new math here?

BeksOmega replied:
The bumpObjects function needs a pretty uncommon set of metrics to work properly:

A) It needs everything in workspace units rather than pixel units.
B) It needs the un-bounded size of the content, even if the content is bounded.

Because the needs were so specific to the bumpObjects function, I wrote everything inside inject, but you're right it is more of a workspace thing. (Also I couldn't come up with a good name, getMetricsUnboundedWorkspaceUnits is a bit longwinded)

I can:
A) Leave as is.
B) Add a new function to the workspace_svg, but I would appreciate a name suggestion.
C) Change the signature of getTopLevelWorkspaceMetrics_() to (bool opt_workspaceUnits, bool opt_forceUnbounded), although this might be a bit confusing because I believe the signature of getMetrics would change for flyouts vs workspaces.

rachel-fenichel replied:
Leave as-is for this change, and add a TODO + github isue to circle back and see if we can move most of it out of inject.

BeksOmega commented on core/workspace_svg.js:
```diff
@@ -1702,10 +1853,11 @@ Blockly.WorkspaceSvg.prototype.setScale = function(newScale) {
   if (this.grid_) {
     this.grid_.update(this.scale);
   }
+  this.scroll_(this.scrollX, this.scrollY);
```
We call scroll_ instead of scrollbar.resize() so that we can center the zoom correctly without scrollbars, but scroll_ does not resize the scrollbars so we have to call resizeContent as well.

Also added the above as a comment.

rachel-fenichel commented on core/inject.js:
```diff
@@ -232,71 +232,163 @@ Blockly.createMainWorkspace_ = function(svg, options, blockDragSurface,
+
+      return workspaceMetrics;
+    };
+    var getWorkspaceObjectMetrics = function(object) {
```
This is a useful function to define directly on blocks and comments, instead of as a helper function that lives inside the inject code.  I'm pretty sure that we're doing exactly this in other places in the code as well, and we can transition those to use the new function as we find them.  It'll be minor code duplication, but it'll make this function easier to find and use (and annotate so it shows up in our documentation).

BeksOmega replied:
Turns out it was already a function: getBoundingRectangle ([block](https://developers.google.com/blockly/reference/js/Blockly.BlockSvg#getBoundingRectangle), [comment](https://developers.google.com/blockly/reference/js/Blockly.WorkspaceCommentSvg#getBoundingRectangle)) the more you know ¯\\\_(ツ)_/¯

BeksOmega commented on core/inject.js:
```diff
@@ -232,71 +232,163 @@ Blockly.createMainWorkspace_ = function(svg, options, blockDragSurface,
-              movedBlocks = true;
-            }
+
+          switch (e.type) {
```
Should a getObjectById() function be added to the workspace so that we can remove this switch statement? Or do you think that's unnecessary?

rachel-fenichel replied:
I think that's unnecessary--blocks and comments share a few APIs, but not many, so you'd still usually have to check the type of the object that you got back.

BeksOmega commented on core/workspace_svg.js:
```diff
@@ -1714,6 +1866,57 @@ Blockly.WorkspaceSvg.prototype.setScale = function(newScale) {
+  this.scrollX = x;
+  this.scrollY = y;
+  if (this.scrollbar) {
+    // The negative x/y values are our new viewLeft and viewTop values.
```
About 2, I changed it over scrollbar.set() but that causes two minor problems:

1. set() calls onScroll_ which calls setMetrics etc, adding unnecessary operations.
2. set() causes a little bit of jitter after zoom: (watch the lower right corner to observe how the scrollbars aren't placed correctly, then observe the upper left corner to observe the jitter)
![scrollbarsset](https://user-images.githubusercontent.com/25440652/52823673-a8501d00-306a-11e9-8b74-c2789fe9a611.gif)

The code is currently calling set(), but I can change it back if those problems are deal breakers.

rachel-fenichel replied:
The setMetrics part wouldn't be a dealbreaker, only becuase I intend to make another pass to sort out metrics stuff.  But the jitter is a problem.  Go ahead and put it back to directly accessing the internals, but add a TODO and a github issue to try to sort those two out and not use the private properties in the future.

BeksOmega commented:
Ok I have done 2000lbs of testing and it looks solid.

Here's everything I tested:

- Zooming with Buttons
  - Movable Workspaces
  - NonMovable Workspaces
- Zooming with Wheel
  - With Scrollbars
  - Without Scrollbars
- Scrolling with Wheel
  - With Scrollbars
  - Without Scrollbars
  - After Zooming in with Buttons
  - After Zooming out with Buttons
  - After Zooming in with Wheel
  - After Zooming out with Wheel
- Dragging the Workspace
  - With Scrollbars
  - Without Scrollbars
  - After Zooming in with Buttons
  - After Zooming out with Buttons
  - After Zooming in with Wheel
  - After Zooming out with Wheel
- Scrolling with Scrollbars
  - After Zooming in with Buttons
  - After Zooming out with Buttons
  - After Zooming in with Wheel
  - After Zooming out with Wheel
- Scroll Center
  - With Scrollbars
  - Without Scrollbars
  - After Zooming in
  - After Zooming out
- Zoom to Fit
  - With Scrollbars
  - Without Scrollbars
- Center on Block
  - With Scrollbars
  - Without Scrollbars
  - After Zooming in
  - After Zooming out
- Block Bumping
  - After Zooming in
  - After Zooming out
- Comment Bumping
  - After Zooming in
  - After Zooming out

And for every one of those I tested every toolbox position w/ categories and w/ simple toolboxes.

I just need replies to [this](https://github.com/google/blockly/pull/2247#discussion_r257051620), [this](https://github.com/google/blockly/pull/2247#discussion_r256634318), and [this](https://github.com/google/blockly/pull/2247#discussion_r256636907), then it should be good to go!

rachel-fenichel commented:
Looks good!  I replied to the outstanding questions; after you handle those I can merge it.

For a few of these my answer is "we should fix it, but not in this PR."  We have until the next release to make sure everything is 100% correct, and those follow-ups are more about code structure than functionality, and can definitely follow in separate PRs.
